GS-hw: fix Burnout games black sky.

iMineLink opened the pull request:
Added an OI hack to perform some point list (CLUT) and sprite (~~CLUT,~~ sky tex) draws via CPU directly in GS local memory, tested only on Burnout 3.
Solves the black sky issue also when TC is cleared.
Update: added on OO hack to download the cloud texture from GPU after rendering during level loading, which solves the missing clouds issue. Tested only in Burnout 3 PAL (50Hz and 60Hz), might need adjustments for different games ~~/regions~~ .
~~Update 2: added a CU hack to avoid upscaling of the clouds drawn during level loading.~~
Update 3: Fixed a bug in TC which assigned a wrong scale factor to native resolution draws.
Update 4: Fixed a bug in the HW renderer which provided too small RT to fit the current draw (upscaled clouds).
Update 5: Reworked and refactored the GS device surface fetching code to always clear the fetched rt or ds.
Update 6: Do not linearly interpolate data when downloading from GPU. Greatly improves clouds rendering with even (x2, x4...) upscaling factors.
~~Sometimes the clouds work, sometimes not.~~
~~Possibly this introduces issues when loading screen is shown multiple times, for example when starting a race, exiting, and starting again.~~
~~Sometimes this intereferes with the skipdraw used to remove the display of the unsupported texture shuffle effect.~~
Testing needed and welcome!

jeremybris commented:
If you still need testers, I'll test this. Idk how though.

iMineLink commented:
Testing is still needed here, and much appreciated, thanks!
To test this PR the simplest way would be to play with Burnout 3 Takedown / Burnout Revenge / Burnout Dominator using the build from this PR artifacts (downloadable from the "Checks" tab) and check the game behaviour, in terms of sky+clouds rendering and visual glitchs/regressions.
I now rebased and force-pushed this PR on top of master, so the artifacts that were previously expired, should be downloadable as soon as the builds are ready.

jeremybris commented:
Works good for me. Thank you for your work. I just noticed one regression: when I switch to "Software" mode, the emulator crashes. There is a small glitch in all of the games, where the clouds are not present, but it's still better than nothing.

| PR: | PS2: |
| ------------- | ------------- |
| ![Capture2](https://user-images.githubusercontent.com/11889359/113067226-f6143a80-9189-11eb-9551-2062168461ca.PNG) | ![Capture16](https://user-images.githubusercontent.com/11889359/113067735-e3e6cc00-918a-11eb-835c-5242a4ce39eb.PNG) |
| ![Capture5](https://user-images.githubusercontent.com/11889359/113067232-f7456780-9189-11eb-9ffe-ac53f25cec50.PNG) |  |
| ![InkedCapture14_LI](https://user-images.githubusercontent.com/11889359/113067551-85215280-918a-11eb-8761-16c8da7eb703.jpg) |  |

MrCK1 commented:
> Works good for me. Thank you for your work. I just noticed one regression: when I switch to "Software" mode, the emulator crashes.

Soudns like you downloaded the 64-bit artifact, SW mode does not work on it at this time. The 32 bit artifact should be fine.

legend800 commented:
That's exactly what I mentioned in the other CR months ago. It's not fixed - the sky is broken, just blue now. No clouds, weird gradient, switching to Software doesn't work/crashes.

JordanTheToaster commented:
Played through 5 races on Burnout 3 and didn't notice anything all looks good to me.

JordanTheToaster commented:
I take that back I've found 2 issues one there seems to be a giant seam in the sky and the lighting seems to be going through the terrain and objects again though I'm not 100% on that second one.

![9764393270275735552_20210331010005_1](https://user-images.githubusercontent.com/80843560/113071486-b9ae0200-91bc-11eb-996a-f93543f843e0.png)

![9764393270275735552_20210331010024_1](https://user-images.githubusercontent.com/80843560/113071499-c16da680-91bc-11eb-91d3-d2f4408174fe.png)

JordanTheToaster commented:
I can reproduce the seam in the sky booting the game normally the sky renders correctly but if I switch to progressive 480p SDTV the seam appears in the sky.

iMineLink commented:
@legend800 could you confirm that this patch still has issues in going back and forth from software mode, specifically the sky going black?

@JordanTheToaster definitely the clouds are not rendered correctly, as also shown by @jeremybris; On that more research should be done as the effect is not straightforward and possibly also the SW implementation, especially that of the point list rendering (for the cloud palette and other palettes), is bugged.

Thanks all!

iMineLink commented:
Thanks @legend800 for the extensive testing.
Clouds rendering still needs improvements here, but we are closer to a working solution with respect to the previous attempt.
If you stick in HW and go back and forth from main menu to races (same or different track), do you see any strange additional behaviour occurring? (Sky and clouds are drawn again via patched code when loading track, I'm not entirely sure that the texture overwriting is ok).

jeremybris commented:
@JordanTheToaster the sun "going through terrain", is already present, #883 , I think.

JordanTheToaster commented:
> @JordanTheToaster the sun "going through terrain", is already present, #883 , I think.

I'm just being forgetful don't worry about that then.

legend800 commented:
> If you stick in HW and go back and forth from main menu to races (same or different track), do you see any strange additional behaviour occurring? (Sky and clouds are drawn again via patched code when loading track, I'm not entirely sure that the texture overwriting is ok).

- Restarting stage is fine
- Leaving and reentering stage is broken though (no clouds)

![2021-03-31 20_06_32-GSdx-hw_ fix Burnout games black sky  by iMineLink · Pull Request #4094 · PCSX2_](https://user-images.githubusercontent.com/20983844/113238558-dfd7ae80-925d-11eb-991b-77a67a4c8702.png)

Loading a different stage is fine and fixes previous stage:
![2021-03-31 20_07_45-Slot_ 0 _ Speed_  82% (49 35) _ Limiter_ Normal _ OGL _ 3200x2240 _ Automatic _ ](https://user-images.githubusercontent.com/20983844/113238668-1281a700-925e-11eb-8ca6-85e76f74e4c9.png)

So so long as you keep playing different stages, it's fine (minus the "seams").

lightningterror commented on pcsx2/GS/Renderers/HW/GSRendererHW.cpp:
```diff
@@ -949,31 +959,16 @@ void GSRendererHW::SwSpriteRender()
+				const GSVector4i A = alpha_a == 0 ? sc : alpha_a == 1 ? dc0 : GSVector4i::zero();
+				const GSVector4i B = alpha_b == 0 ? sc : alpha_b == 1 ? dc0 : GSVector4i::zero();
+				const GSVector4i C = alpha_c == 2 ? GSVector4i(alpha_fix).xxxx().ps32() :
+					(alpha_c == 0 ? sc : dc0)
```
I'd move this above, it's not that big.

RedDevilus commented:
Might be best to rename the commits to `GS` instead of `GSdx`

RedDevilus commented:
Can't recall if the clouds are supposed to cut-out so abruptly, should check on my console to reaffirm.
![gs_20211005021526](https://user-images.githubusercontent.com/24227051/135941388-20ca7a95-63fa-405e-a985-c1afbd13e1ba.png)
[gs_20211005021526.zip](https://github.com/PCSX2/pcsx2/files/7281964/gs_20211005021526.zip)

iMineLink commented:
The clouds are bugged but I have to find the offending draw call(s). I feel that the bug is related to something wrong in the point list palette draw hack.

iMineLink commented:
The point list palette draw hack had blending completely broken.
I attempted to fix this by recycling the same rendering code used for the sw sprite renderer.
Gs dumps are mostly useless here.
The issue with gs dumps is that the sky related draws (hacked here) happen during level loading, so you do need a huge multiframe dump from the beginning of the loading to debug them and their result on the sky appearance.
Anyway, please go ahead and give this another testing round!

JordanTheToaster commented:
The giant seam in the sky in Burnout 3 is gone but the clouds are now gone which is a problem.

![unknown](https://user-images.githubusercontent.com/80843560/140591343-4f109821-e1e7-42f0-95dd-c78692cf45c4.png)

iMineLink commented:
There's more render to clut than expected here.
I'll try plugging in the true sw renderer to perform those draws.

iMineLink commented:
Not so easy to plug in the sw renderer, the rasterizer unfortunately requires quite a big setup to perform a draw which ultimately seems feasible to mantain only inside a `GSRendererSW` instance.
Anyway, I triple checked, and for the blending that was performed by the clouds palette draw (A = B = 2, C = 2, D = 0), without alpha correction and alpha/depth tests, it is ok to copy the RGBA color as is from the vertex to the framebuffer; So blending was broken in general but not for this specific case.
Instead I realized that the hack worked on top of [9832a935d7416fb06efe78922b55417a39949ae1](https://github.com/PCSX2/pcsx2/commit/9832a935d7416fb06efe78922b55417a39949ae1) (~3 months ago), but the same does not work on top of current master (no clouds).
I didn't try to rebase on top of intermediate commits though, but clearly something really subtle is going on; I'll continue the research.

JordanTheToaster commented:
Still no clouds sadly we'll get there one day.

iMineLink commented:
Please test now.

Clouds should work in Burnout 3 PAL, for other games I will need some help in retrieving the 32 bit tex copy draw call during level loading (later the clouds tex is rendered from the copy on top of the sky using a palette at each frame).

JordanTheToaster commented:
The color of the sky is correct but clouds are still missing from Burnout 3 NTSC.

![image](https://user-images.githubusercontent.com/80843560/142764932-6e0e18ac-87f8-42b9-aab0-f5bddd25c89d.png)

iMineLink commented:
I attempted at generalizing the cloud rendering detection during level loading, and the code is now working for B3 PAL 50Hz and 60Hz (is it the same at NTSC?).
I wrongly supposed it was a 32 bit texture copy, the draw is instead a proper rendering with a palette.

JordanTheToaster commented:
Clouds now somewhat showup correctly though they appear to be full of holes and the giant seam is also back.

![image](https://user-images.githubusercontent.com/80843560/142778290-65b56eb1-17e4-466b-b381-74b9cccfe919.png)

lightningterror commented on pcsx2/GS/GSVector4i.h:
```diff
@@ -2203,4 +2206,11 @@ class alignas(16) GSVector4i
 	__forceinline static GSVector4i xff(int n) { return m_xff[n]; }
 	__forceinline static GSVector4i x0f(int n) { return m_x0f[n]; }
+
+	std::string to_string_xyzw()
```
Maybe not worth it for a log to include a libraries.

iMineLink commented:
The buggy clouds ("holes" + "seam") are an upscaling only issue now.
With the last force push I fixed the "holes" by avoiding the upscale of the cloud texture draw during level loading (the one which is downloaded forcibly); Later it is used as an indexed texture at the beginning of each frame to draw the clouds on top of the sky so it's crucial to have it precisely drawn the first time.
Now only the "seam" remain, still an upscaling only issue, I think due to an offset introduced in the UV coordinates of the initial draw when there is upscaling, which is being investigated (I can already detect the draw luckily).
I hope that's the case.

iMineLink commented:
I must be honest, this was a tough one. There was a bug in the TC (....really??) which did not reset the scale factor to 1 when an old texture was reused as curent RT for a draw at forced native resolution (the clouds). So only a portion of the clouds were downloaded because the wrong (non 1) scale factor was chopping the area to be copied. Hence the "seam" in the clouds.

It should now be fixed.

There's only one thing that sometimes occurs when I load the game but I do not know if it is related to this PR or not: the yellow stripes, which the other dedicated CRC hack `GSC_BurnoutGames` should skip drawing, instead appear on screen.

If testing demonstrates that there are no regressions, I think we should be good to go (code review aside)!

iMineLink commented:
A fix for a bug in the HW renderer made also the clouds drawn at upscaled resolution possible (the RT was too small before).

RedDevilus commented:
Burnout Dominator works, needs testing of other TC-buggy games.

refractionpcsx2 commented:
A kind of win/loss here

This PR introduces bars at the bottom of the screen in test drive, however it kinda breaks and half fixes it when you upscale.  Here's a dump for you to look at.
[testdrive.zip](https://github.com/PCSX2/pcsx2/files/7614835/testdrive.zip)

JordanTheToaster commented:
Burnout 3 NTSC works perfectly now upscaled and haven't noticed any other issues with the games I currently own so it all looks good to me.

iMineLink commented:
I reverted the automatic rt/ds growth to match the current draw size because it introduces regressions in Test Drive (it makes an unsupported effect, previously luckily cut off, appear).
So I re-added the CU hack to avoid clouds upscale (1024 * 256 draw at native res), which were going to be downloaded (so downscaled to native) in any case.
I also removed the code which allowed dumping of the alpha channel for all the textures when debugging.
Both the rt/ds automatic sizing and the alpha channel dumping will be addressed by dedicated PRs.
A small side effect of not having auto rt/ds size is that the reflections on the cars are a bit worse, but this can be fixed later too.

refractionpcsx2 commented:
I'd rather have the regression in Test drive if it's more accurate TC behaviour, I don't want to avoid progress because it uncovers bugs. if a bug shows it's head, tough, we'll have to deal with it.

iMineLink commented:
I also think that the auto rt/ds resizing is a step towards more accurate TC (we should really call it texture manager or surface manager or gs memory proxy cause it's really more than a mere "cache") behaviour.
So I restored auto resizing and removed the CU hack.
As for the alpha channel dump, it is re-added but only for dev-builds (should be fine all the time).

This way, by merging this we regress Test Drive in HW, but it can be fixed later.

refractionpcsx2 commented:
What is the alpha dumping going to do in dev builds? I presume nothing unless dumping is enabled for that draw?

lightningterror commented:
> I'd rather have the regression in Test drive if it's more accurate TC behaviour, I don't want to avoid progress because it uncovers bugs. if a bug shows it's head, tough, we'll have to deal with it.

Agreed.

iMineLink commented:
@refractionpcsx2 yeah pretty much. It says that when a gs texture is saved to png, also a grayscale png with the alpha content must be saved. Even though it's pretty buggy this way ("alpha" is stored also when there isn't any in the format), it's better than nothing.

ribolovac510 commented:
Black (SLUS-21376) gets stretched out if upscaled on this PR and the black sky is still present if you don't switch to Software rendering during loading.

| PR | v1.7.2092 |
| :---: | :---: |
| 4x Native | 4x Native + F9 during loading |
| ![gs_20211203100118](https://user-images.githubusercontent.com/74596991/144576051-d7150623-5aed-4f1e-92da-a747052accf1.png) | ![gs_20211203095933](https://user-images.githubusercontent.com/74596991/144576167-72ee39cc-4c7f-480b-804c-1dbd0bd9c31d.png) |
| Native |  |
| ![gs_20211203100137](https://user-images.githubusercontent.com/74596991/144576309-2554486b-022a-4854-aa2c-fe75fd3c82b4.png) |  |

Tested it with #3895 also, the image does not get stretched out on that PR but the sky is still black without switching.

iMineLink commented:
The stretching might be a regression, I'll look into it.
Don't expect Black sky to be not black anymore from this PR, it's only addressing Burnout games.
Later with small effort it can be extended to Black.

iMineLink commented:
I fixed the auto rt/ds resizing code, Black looks ok also upscaled now.

DonelBueno commented:
Now Burnout games (at least Burnou 3: Takedown) have issues when upscaled:
![PCSX2_2021_12_03_19_40_53_887](https://user-images.githubusercontent.com/4239694/144682366-252dcbbd-1872-4527-af4c-74101ef3f760.png)

RedDevilus commented:
What's the issue exactly with that screenshot? That the clouds don't look sharp?

refractionpcsx2 commented:
I'm assuming the bottom of the screen is cut off and it looks kinda stretched vertically.

iMineLink commented:
Could not reproduce, anyway I improved a bit TC. Anyway I noticed that above 3x upscaling the clouds are broken ("holes", probably downscaling is too aggressive when downloading from a huge texture).

refractionpcsx2 commented:
Kingdom hearts shadows are interesting when upscaled

![image](https://user-images.githubusercontent.com/6278726/144688309-c0da031c-c9ee-40aa-9b1a-b18c1a83dcde.png)

Test Drive has changed (in any resolution)

![image](https://user-images.githubusercontent.com/6278726/144688348-cd3aa23c-35d3-47c2-bdd7-e1c1669dbd05.png)


Couldn't reproduce any problem with Burnout 3.

iMineLink commented:
Test Drive is fixed now. Can you retest KH or provide a dump?

refractionpcsx2 commented:
Confirmed test drive is now okay, this is the situation with KH now

![image](https://user-images.githubusercontent.com/6278726/144724717-ef144c4e-ac68-49ed-aa4c-4b0f5da563d8.png)

I'll attach the gsdump so you can check yourself
[kingdomhearts.zip](https://github.com/PCSX2/pcsx2/files/7654701/kingdomhearts.zip)

refractionpcsx2 commented:
Got another one for ya :) Simpsons freaks out when upscaling
[simpsons_upscale.zip](https://github.com/PCSX2/pcsx2/files/7654733/simpsons_upscale.zip)

refractionpcsx2 commented:
actually don't worry about the Kingdom Hearts thing, that happens on master when upscaling too.

But the simpsons is a problem

DonelBueno commented:
Now Burnout looks perfect, no stretching or cut off image.

refractionpcsx2 commented:
Ah so it was old data? Nice work :)

iMineLink commented:
It seems so!

B3 clouds only work with odd (x3, x5...) upscaling factors though.
It's a huge 32 bit texture draw (only alpha channel relevant), which gets downloaded (so downscaled) to gs memory.
Later the clouds data is read as 8 bit texture for drawing.
I guess it's some missing half pixel offset somewhere that breaks the rendering with even (x2, x4...) upscaling.

Either we force native resolution draw for the clouds (it is downscaled later in any case), or also this should be fixed.

Again, if we force native res draw, this PR can be split (burnout specific improvements here, improvements to rendering code in another one).

refractionpcsx2 commented:
hmm I don't know why even scaling doesn't work, that's a bit strange, I mean every scale is a multiple so it should scale perfectly :/ Sounds like a bug of some sort, probably worth fixing.  I want to avoid CRC hacks if we can help it.

iMineLink commented:
With even upscaling, the pixel center in the native texture becomes the boundary between two adjacent pixels in the upscaled texture; this does not occur with odd upscaling, because the center will still be in the middle of a pixel.
This resulted, with even upscaling, in a blurred downscaled texture, which made the masked rendering wrong later (the interpolation of the alpha values broke the masking).
Avoiding linear interpolation when downloading (so downscaling) data greatly improves the rendering quality in this case, and looks a better strategy overall.

refractionpcsx2 commented:
sounds good to me :) man I hate the pixel positions on PS2 lol. I wonder if we should round/correct vertecies/UV's/ST's as they come in before they're passed to the renderer and upscaled etc.

lightningterror commented:
@iMineLink looks like a pr caused a bunch of conflicts, can you resolve them, I want this next on the merge list since it's been open for a while.

iMineLink commented:
done

JordanTheToaster commented:
Clouds look low res and the seam is back at any upscaling. Native works correctly.

![image](https://user-images.githubusercontent.com/80843560/145475182-f6a934ff-2302-4647-baf7-502435851ad4.png)

iMineLink commented:
a bug has been fixed but the rendering is broken now. needs time to be checked.

iMineLink commented:
rebased on latest master dropping the last bugfix commit which was already merged. working on my side now.

lightningterror commented:
Some new warnings which should be taken care of
![image](https://user-images.githubusercontent.com/18107717/145716294-7087b636-f946-4aed-8e7c-b61cc5d90e53.png)

JordanTheToaster reviewed (approved):
Works at native 2x 3x ect all fine clouds look perfect and the seam is gone.

![image](https://user-images.githubusercontent.com/80843560/145716565-4544bf85-3a13-4eb1-9b9c-0804c9152836.png)

refractionpcsx2 commented:
Tested a bunch of games and GS dumps, all looks good :)

RedDevilus reviewed (approved):
No more issues :)

Bring back the rainbow palette.

refractionpcsx2 commented:
@iMineLink  I think this is good to merge now, you happy to do so or is there anything else you want to do to it?

iMineLink commented:
For me it's ok to merge!

lightningterror reviewed (approved):
Time for the public to enjoy the changes.

freedom85uk commented:
Hi,

I am new to emulation. Excuse my lack of knowledge, but has this been updated within the source code of PCSX2? I am running the latest nightly build (v1.7.4023). Will this build contain the changes or do I need to something in addition?

Thanks in advance
Luke

refractionpcsx2 commented:
Every new version contains everything from older versions.

If you need further support please use our discord, github is not the place for it

freedom85uk commented:
Thanks. 

Apologies, I didn't realise I was unable to ask a question on here regarding this issue.
